include/h2o/socket.h

```diff
@@ -217,7 +221,7 @@ void h2o_socket_dont_read(h2o_socket_t *sock, int dont_read);
 /**
  * connects to peer
  */
-h2o_socket_t *h2o_socket_connect(h2o_loop_t *loop, struct sockaddr *addr, socklen_t addrlen, h2o_socket_cb cb);
+h2o_socket_t *h2o_socket_connect(h2o_loop_t *loop, struct sockaddr *addr, socklen_t addrlen, h2o_socket_cb cb, int *so_err);
```

Comment:
`destination_ip_unroutable` and `connection_refused` are distinct errors in the proxy-status system, and we are able to distinguish the two case when the connecting socket blocks a little bit before finding out.  It did not seem right that we would not be able to distinguish those two cases just because the kernel reports them immediately without blocking.  But of course we do not need that at all cost.  Can you think of ways to make this API better and still recognize the `destination_ip_unroutable` condition?

I agree that it is bad layering to be returning an errno and that that does not work for libuv.  That is one of the main reasons why the PR is still a draft.  It makes sense to use string pointers to identify errors, as you write in another comment.  Conceptually this function returns either a socket or an error, and in Rust the idiomatic thing to do would be to return a Result type.  Another option in theory would be to invoke the callback in all cases, unless the allocation fails, but that would mean that we add a memory allocation in the error case, and that does not seem right either.  What options can you think of please?

Reply:
Yeah, thinking a bit more, I tend to believe that we might want to bisect between two cases that both return NULL: when a file descriptor is unavailable and when the route is unavailable. The former is a temporary error for which we might want to return 503, while in case of ENETUNREACH the recommended status code is 502.

Based on that observation, I think we have to somehow return `const char *` alongside the return value of NULL. Maybe change `int *so_err` to `const char **err` and call it a day?

Reply:
Done in 84496d135.